File: Aero Linea/Aerolinea.API/Aerolinea.API/Repositories/UsuarioRepository.cs
using Aerolinea.API.Data;
using Aerolinea.API.DTOs;
using Aerolinea.API.Models;
using Microsoft.Data.SqlClient;

namespace Aerolinea.API.Repositories
{
    /// <summary>
    /// Repositorio de usuarios. Gestiona la creacion, consulta y actualizacion de usuarios,
    /// incluyendo la asignacion de nacionalidades, verificacion de duplicados en registro,
    /// busqueda por credenciales para autenticacion y administracion de roles.
    /// </summary>
    public class UsuarioRepository : IUsuarioRepository
    {
        private readonly DbConnectionFactory _connectionFactory;
        private readonly NacionalidadRepository _nacionalidadRepository;
        private readonly CiudadRepository _ciudadRepository;

        public UsuarioRepository(
            DbConnectionFactory connectionFactory,
            NacionalidadRepository nacionalidadRepository,
            CiudadRepository ciudadRepository)
        {
            _connectionFactory = connectionFactory;
            _nacionalidadRepository = nacionalidadRepository;
            _ciudadRepository = ciudadRepository;
        }

        /// <summary>
        /// Inserta un nuevo usuario en la base de datos y retorna el ID generado.
        /// Los campos opcionales como telefono, fecha de nacimiento y ciudad aceptan
        /// null o valor por defecto y se almacenan como DBNull en la base de datos.
        /// </summary>
        public async Task<int> CrearUsuario(Usuario usuario)
        {
            using var connection = _connectionFactory.CreateConnection();
            await connection.OpenAsync();

            var query = @"
                INSERT INTO Usuario (Correo, ContrasenaHash, Pasaporte, Username, Nombre, Apellido,
                                    Telefono, FechaNacimiento, CiudadId, RolID)
                OUTPUT INSERTED.Id
                VALUES (@Correo, @ContrasenaHash, @Pasaporte, @Username, @Nombre, @Apellido,
                        @Telefono, @FechaNacimiento, @CiudadId, @RolID)";

            using var command = new SqlCommand(query, connection);
            command.Parameters.AddWithValue("@Correo", usuario.Correo);
            command.Parameters.AddWithValue("@ContrasenaHash", usuario.ContrasenaHash);
            command.Parameters.AddWithValue("@Pasaporte", usuario.Pasaporte);
            command.Parameters.AddWithValue("@Username", usuario.Username);
            command.Parameters.AddWithValue("@Nombre", usuario.Nombre);
            command.Parameters.AddWithValue("@Apellido", usuario.Apellido);
            command.Parameters.AddWithValue("@Telefono", (object)usuario.Telefono ?? DBNull.Value);
            command.Parameters.AddWithValue("@FechaNacimiento", usuario.FechaNacimiento == DateTime.MinValue ? DBNull.Value : usuario.FechaNacimiento);
            command.Parameters.AddWithValue("@CiudadId", usuario.CiudadId == 0 ? DBNull.Value : usuario.CiudadId);
            command.Parameters.AddWithValue("@RolID", usuario.RolID);

            return (int)await command.ExecuteScalarAsync();
        }

        /// <summary>
        /// Asocia una lista de nacionalidades a un usuario recien creado.
        /// Por cada nacionalidad obtiene o crea el registro en la tabla Nacionalidad
        /// y luego inserta el vinculo en UsuarioNacionalidad.
        /// </summary>
        public async Task AgregarNacionalidades(int usuarioId, List<string> nacionalidades)
        {
            using var connection = _connectionFactory.CreateConnection();
            await connection.OpenAsync();

            foreach (var nacionalidad in nacionalidades)
            {
                int nacionalidadId = await _nacionalidadRepository.ObtenerOCrearId(nacionalidad, connection);

                using var command = new SqlCommand(
                    "INSERT INTO UsuarioNacionalidad (UsuarioId, NacionalidadId) VALUES (@UsuarioId, @NacionalidadId)",
                    connection);
                command.Parameters.AddWithValue("@UsuarioId", usuarioId);
                command.Parameters.AddWithValue("@NacionalidadId", nacionalidadId);
                await command.ExecuteNonQueryAsync();
            }
        }

        /// <summary>
        /// Verifica si ya existe un usuario con el correo, username o pasaporte indicados.
        /// Retorna un objeto RegisterConstraint con las flags correspondientes a cada campo
        /// que ya este en uso, para informar al cliente que datos causan conflicto.
        /// </summary>
        public async Task<RegisterConstraint> VerificarExistencia(string correo, string username, string pasaporte)
        {
            using var connection = _connectionFactory.CreateConnection();
            await connection.OpenAsync();

            var constraint = new RegisterConstraint();

            using var command = new SqlCommand(
                "SELECT COUNT(*) FROM Usuario WHERE Correo = @Correo", connection);
            command.Parameters.AddWithValue("@Correo", correo);
            constraint.CorreoExiste = (int)await command.ExecuteScalarAsync() > 0;

            command.CommandText = "SELECT COUNT(*) FROM Usuario WHERE Username = @Username";
            command.Parameters.Clear();
            command.Parameters.AddWithValue("@Username", username);
            constraint.UsernameExiste = (int)await command.ExecuteScalarAsync() > 0;

            command.CommandText = "SELECT COUNT(*) FROM Usuario WHERE Pasaporte = @Pasaporte";
            command.Parameters.Clear();
            command.Parameters.AddWithValue("@Pasaporte", pasaporte);
            constraint.PasaporteExiste = (int)await command.ExecuteScalarAsync() > 0;

            return constraint;
        }

        /// <summary>
        /// Busca un usuario por correo electronico o username para el proceso de autenticacion.
        /// Retorna el objeto Usuario completo con su hash de contrasena y rol, o null si
        /// no existe ningun usuario con esas credenciales.
        /// </summary>
        public async Task<Usuario?> ObtenerPorCorreoOUsername(string correoOUsername)
        {
            using var connection = _connectionFactory.CreateConnection();
            await connection.OpenAsync();

            var query = @"
                SELECT Id, Correo, ContrasenaHash, Pasaporte, Username, Nombre, Apellido,
                       Telefono, FechaNacimiento, CiudadId, RolID
                FROM Usuario
                WHERE Correo = @CorreoOUsername OR Username = @CorreoOUsername";

            using var command = new SqlCommand(query, connection);
            command.Parameters.AddWithValue("@CorreoOUsername", correoOUsername);

            using var reader = await command.ExecuteReaderAsync();

            if (await reader.ReadAsync())
            {
                return new Usuario
                {
                    Id = reader.GetInt32(0),
                    Correo = reader.GetString(1),
                    ContrasenaHash = reader.GetString(2),
                    Pasaporte = reader.IsDBNull(3) ? "" : reader.GetString(3),
                    Username = reader.IsDBNull(4) ? "" : reader.GetString(4),
                    Nombre = reader.GetString(5),
                    Apellido = reader.GetString(6),
                    Telefono = reader.IsDBNull(7) ? "" : reader.GetString(7),
                    FechaNacimiento = reader.IsDBNull(8) ? DateTime.MinValue : reader.GetDateTime(8),
                    CiudadId = reader.IsDBNull(9) ? 0 : reader.GetInt32(9),
                    RolID = reader.GetInt32(10)
                };
            }

            return null;
        }

        /// <summary>
        /// Retorna el nombre del rol asociado al ID indicado.
        /// Retorna null si el rol no existe en la tabla Rol.
        /// </summary>
        public async Task<string?> ObtenerNombreRol(int rolId)
        {
            using var connection = _connectionFactory.CreateConnection();
            await connection.OpenAsync();

            using var command = new SqlCommand(
                "SELECT RolNombre FROM Rol WHERE ID = @RolId", connection);
            command.Parameters.AddWithValue("@RolId", rolId);

            var result = await command.ExecuteScalarAsync();
            return result?.ToString();
        }

        /// <summary>
        /// Actualiza el rol de un usuario al nuevo rol indicado.
        /// Retorna true si la actualizacion afecto al menos una fila.
        /// </summary>
        public async Task<bool> ActualizarRol(int usuarioId, int nuevoRolId)
        {
            using var connection = _connectionFactory.CreateConnection();
            await connection.OpenAsync();

            var query = "UPDATE Usuario SET RolID = @NuevoRolId WHERE Id = @UsuarioId";

            using var command = new SqlCommand(query, connection);
            command.Parameters.AddWithValue("@UsuarioId", usuarioId);
            command.Parameters.AddWithValue("@NuevoRolId", nuevoRolId);

            return await command.ExecuteNonQueryAsync() > 0;
        }

        /// <summary>
        /// Verifica si existe un usuario con el ID indicado.
        /// Retorna true si se encuentra al menos un registro con ese ID.
        /// </summary>
        public async Task<bool> UsuarioExiste(int usuarioId)
        {
            using var connection = _connectionFactory.CreateConnection();
            await connection.OpenAsync();

            using var command = new SqlCommand(
                "SELECT COUNT(*) FROM Usuario WHERE Id = @UsuarioId", connection);
            command.Parameters.AddWithValue("@UsuarioId", usuarioId);

            return (int)await command.ExecuteScalarAsync() > 0;
        }

        /// <summary>
        /// Verifica si existe un rol con el ID indicado en la tabla Rol.
        /// Retorna true si se encuentra al menos un registro con ese ID.
        /// </summary>
        public async Task<bool> RolExiste(int rolId)
        {
            using var connection = _connectionFactory.CreateConnection();
            await connection.OpenAsync();

            using var command = new SqlCommand(
                "SELECT COUNT(*) FROM Rol WHERE Id = @RolId", connection);
            command.Parameters.AddWithValue("@RolId", rolId);

            return (int)await command.ExecuteScalarAsync() > 0;
        }

        /// <summary>
        /// Retorna la lista completa de usuarios registrados en el sistema, ordenados por ID.
        /// Incluye todos los campos del perfil pero no incluye las nacionalidades asociadas.
        /// </summary>
        public async Task<List<Usuario>> ObtenerTodos()
        {
            using var connection = _connectionFactory.CreateConnection();
            await connection.OpenAsync();

            var lista = new List<Usuario>();
            using var command = new SqlCommand(
                "SELECT Id, Correo, ContrasenaHash, Pasaporte, Username, Nombre, Apellido, Telefono, FechaNacimiento, CiudadId, RolID FROM Usuario ORDER BY Id",
                connection);
            using var reader = await command.ExecuteReaderAsync();

            while (await reader.ReadAsync())
            {
                lista.Add(new Usuario
                {
                    Id = reader.GetInt32(0),
                    Correo = reader.GetString(1),
                    ContrasenaHash = reader.GetString(2),
                    Pasaporte = reader.IsDBNull(3) ? "" : reader.GetString(3),
                    Username = reader.IsDBNull(4) ? "" : reader.GetString(4),
                    Nombre = reader.GetString(5),
                    Apellido = reader.GetString(6),
                    Telefono = reader.IsDBNull(7) ? "" : reader.GetString(7),
                    FechaNacimiento = reader.IsDBNull(8) ? DateTime.MinValue : reader.GetDateTime(8),
                    CiudadId = reader.IsDBNull(9) ? 0 : reader.GetInt32(9),
                    RolID = reader.GetInt32(10)
                });
            }

            return lista;
        }
    }
}

File: Cadena de Hoteles/Hotel/src/main/java/org/example/repositories/UsuarioRepository.java
package org.example.repositories;

import org.example.data.DataAccessException;
import org.example.data.DatabaseManager;
import org.example.dtos.UsuarioAdminDTO;
import org.example.dtos.UsuarioPerfilResponseDTO;
import org.example.dtos.UsuarioValidacionResponseDTO;
import org.example.helpers.CamposDuplicadosException;

import java.util.List;

/**
 * Repository para el acceso a datos de usuarios.
 * Cubre validaciones de unicidad, creacion, consulta de perfil,
 * gestion de contrasenas y operaciones administrativas de rol.
 */
public class UsuarioRepository {

    // Validaciones de unicidad antes de registrar un usuario

    /**
     * Verifica si ya existe un usuario con el username dado.
     * @param username nombre de usuario a verificar.
     * @return true si el username ya esta en uso, false en caso contrario.
     */
    public boolean existeUsername(String username) {
        String sql = "SELECT COUNT(*) FROM Usuario WHERE Username = ?";
        List<Integer> result = DatabaseManager.executeQuery(sql, rs -> rs.getInt(1), username);
        return !result.isEmpty() && result.get(0) > 0;
    }

    /** Verifica si el username está en uso por un usuario diferente al indicado. */
    public boolean existeUsernameExceptoId(String username, int excluirId) {
        String sql = "SELECT COUNT(*) FROM Usuario WHERE Username = ? AND ID != ?";
        List<Integer> result = DatabaseManager.executeQuery(sql, rs -> rs.getInt(1), username, excluirId);
        return !result.isEmpty() && result.get(0) > 0;
    }

    /** Verifica si el correo está en uso por un usuario diferente al indicado. */
    public boolean existeCorreoExceptoId(String correo, int excluirId) {
        String sql = "SELECT COUNT(*) FROM Usuario WHERE Correo = ? AND ID != ?";
        List<Integer> result = DatabaseManager.executeQuery(sql, rs -> rs.getInt(1), correo, excluirId);
        return !result.isEmpty() && result.get(0) > 0;
    }

    /** Verifica si el pasaporte está en uso por un usuario diferente al indicado. */
    public boolean existePasaporteExceptoId(String pasaporte, int excluirId) {
        if (pasaporte == null || pasaporte.isBlank()) return false;
        String sql = "SELECT COUNT(*) FROM Usuario WHERE Pasaporte = ? AND ID != ?";
        List<Integer> result = DatabaseManager.executeQuery(sql, rs -> rs.getInt(1), pasaporte, excluirId);
        return !result.isEmpty() && result.get(0) > 0;
    }

    /**
     * Verifica si ya existe un usuario registrado con el correo dado.
     * @param correo correo electronico a verificar.
     * @return true si el correo ya esta en uso, false en caso contrario.
     */
    public boolean existeCorreo(String correo) {
        String sql = "SELECT COUNT(*) FROM Usuario WHERE Correo = ?";
        List<Integer> result = DatabaseManager.executeQuery(sql, rs -> rs.getInt(1), correo);
        return !result.isEmpty() && result.get(0) > 0;
    }

    /**
     * Verifica si ya existe un usuario registrado con el numero de pasaporte dado.
     * Retorna false directamente si el pasaporte es nulo o esta vacio.
     * @param pasaporte numero de pasaporte a verificar.
     * @return true si el pasaporte ya esta en uso, false en caso contrario.
     */
    public boolean existePasaporte(String pasaporte) {
        if (pasaporte == null || pasaporte.isBlank()) return false;
        String sql = "SELECT COUNT(*) FROM Usuario WHERE Pasaporte = ?";
        List<Integer> result = DatabaseManager.executeQuery(sql, rs -> rs.getInt(1), pasaporte);
        return !result.isEmpty() && result.get(0) > 0;
    }

    // Creacion de nuevos usuarios

    /**
     * Inserta un nuevo usuario en la base de datos con rol de cliente por defecto.
     * Si Oracle lanza ORA-00001 (restriccion unica violada) por una race condition,
     * re-verifica que campos estan duplicados y lanza {@link CamposDuplicadosException}.
     * @param correo              correo electronico del usuario.
     * @param contrasenaHasheada  contrasena ya procesada con hash.
     * @param pasaporte           numero de pasaporte del usuario, puede ser nulo.
     * @param username            nombre de usuario unico.
     * @param nombre              nombre de pila del usuario.
     * @param apellido            apellido del usuario.
     * @param telefono            numero de telefono de contacto.
     * @param fechaNacimiento     fecha de nacimiento del usuario.
     * @param ciudadId            ID de la ciudad de residencia del usuario.
     * @return ID generado por la base de datos para el nuevo usuario.
     * @throws CamposDuplicadosException si correo, pasaporte o username ya existen.
     */
    public int crearUsuario(
            String correo,
            String contrasenaHasheada,
            String pasaporte,
            String username,
            String nombre,
            String apellido,
            String telefono,
            java.sql.Date fechaNacimiento,
            int ciudadId,
            String preferenciasOferta
    ) {
        String sql = """
                INSERT INTO Usuario
                    (Correo, Contrasena, Pasaporte, Username, Nombre, Apellido,
                     Rol_ID, Telefono, Fecha_Nacimiento, Ciudad_ID, Preferencias_Oferta)
                VALUES (?, ?, ?, ?, ?, ?, 1, ?, ?, ?, ?)
                """;

        try {
            return DatabaseManager.executeInsertReturnId(
                    sql, "ID",
                    correo, contrasenaHasheada, pasaporte, username,
                    nombre, apellido, telefono, fechaNacimiento, ciudadId, preferenciasOferta
            );
        } catch (DataAccessException e) {
            Throwable cause = e.getCause();
            if (cause instanceof java.sql.SQLException sqlEx && sqlEx.getErrorCode() == 1) {
                UsuarioValidacionResponseDTO campos = new UsuarioValidacionResponseDTO(
                        existeUsername(username),
                        existeCorreo(correo),
                        existePasaporte(pasaporte)
                );
                throw new CamposDuplicadosException(campos);
            }
            throw e;
        }
    }

    // Consulta del perfil completo del usuario

    /**
     * Retorna el perfil completo de un usuario incluyendo su ciudad y pais.
     * @param usuarioId ID del usuario a consultar.
     * @return DTO con los datos del perfil, o null si el usuario no existe.
     */
    public UsuarioPerfilResponseDTO obtenerPerfil(int usuarioId) {
        String sql = """
                SELECT u.ID, u.Username, u.Correo, u.Pasaporte, u.Nombre, u.Apellido,
                       u.Telefono, u.Fecha_Nacimiento, u.Rol_ID,
                       u.Preferencias_Oferta,
                       c.Nombre AS Ciudad, p.Nombre AS Pais
                FROM   Usuario u
                LEFT JOIN Ciudad c ON u.Ciudad_ID = c.ID
                LEFT JOIN Pais   p ON c.Pais_ID   = p.ID
                WHERE  u.ID = ?
                """;

        // Mapea la fila del resultado al DTO de perfil, manejando fecha nula
        List<UsuarioPerfilResponseDTO> result = DatabaseManager.executeQuery(sql, rs -> {
            UsuarioPerfilResponseDTO dto = new UsuarioPerfilResponseDTO();
            dto.setId(rs.getInt("ID"));
            dto.setUsername(rs.getString("Username"));
            dto.setCorreo(rs.getString("Correo"));
            dto.setPasaporte(rs.getString("Pasaporte"));
            dto.setNombre(rs.getString("Nombre"));
            dto.setApellido(rs.getString("Apellido"));
            dto.setTelefono(rs.getString("Telefono"));

            java.sql.Date fecha = rs.getDate("Fecha_Nacimiento");
            dto.setFechaNacimiento(fecha != null ? fecha.toString() : null);

            dto.setRolId(rs.getInt("Rol_ID"));
            dto.setCiudad(rs.getString("Ciudad"));
            dto.setPais(rs.getString("Pais"));
            dto.setPreferenciasOferta(rs.getString("Preferencias_Oferta"));
            return dto;
        }, usuarioId);

        return result.isEmpty() ? null : result.get(0);
    }

    // Nacionalidades asociadas al usuario

    /**
     * Retorna los nombres de las nacionalidades registradas para un usuario.
     * @param usuarioId ID del usuario a consultar.
     * @return lista de nombres de nacionalidades del usuario.
     */
    public List<String> obtenerNacionalidades(int usuarioId) {
        String sql = """
                SELECT n.Nombre
                FROM   UsuarioNacionalidad un
                JOIN   Nacionalidad n ON un.Nacionalidad_ID = n.ID
                WHERE  un.Usuario_ID = ?
                """;

        return DatabaseManager.executeQuery(sql, rs -> rs.getString("Nombre"), usuarioId);
    }

    // Actualizacion de datos de contacto

    /**
     * Actualiza el numero de telefono de un usuario.
     * @param usuarioId ID del usuario a modificar.
     * @param telefono  nuevo numero de telefono.
     */
    public void actualizarTelefono(int usuarioId, String telefono) {
        String sql = "UPDATE Usuario SET Telefono = ? WHERE ID = ?";
        DatabaseManager.executeUpdate(sql, telefono, usuarioId);
    }

    // Gestion de contrasena

    /**
     * Retorna la contrasena hasheada almacenada para un usuario.
     * @param usuarioId ID del usuario a consultar.
     * @return contrasena hasheada, o null si el usuario no existe.
     */
    public String obtenerContrasena(int usuarioId) {
        String sql = "SELECT Contrasena FROM Usuario WHERE ID = ?";
        List<String> result = DatabaseManager.executeQuery(sql, rs -> rs.getString("Contrasena"), usuarioId);
        return result.isEmpty() ? null : result.get(0);
    }

    /**
     * Actualiza la contrasena de un usuario con el nuevo hash proporcionado.
     * @param usuarioId          ID del usuario a modificar.
     * @param contrasenaHasheada nueva contrasena ya procesada con hash.
     */
    public void actualizarContrasena(int usuarioId, String contrasenaHasheada) {
        String sql = "UPDATE Usuario SET Contrasena = ? WHERE ID = ?";
        DatabaseManager.executeUpdate(sql, contrasenaHasheada, usuarioId);
    }

    // Operaciones administrativas

    /**
     * Retorna todos los usuarios del sistema con su rol, ciudad y pais asignados.
     * @return lista de DTOs con la informacion administrativa de cada usuario.
     */
    public List<UsuarioAdminDTO> listarTodosConRol() {
        String sql = """
                SELECT u.ID, u.Username, u.Nombre, u.Apellido, u.Correo, u.Telefono,
                       u.Fecha_Nacimiento, u.Rol_ID,
                       r.RolNombre,
                       c.Nombre AS Ciudad,
                       p.Nombre AS Pais
                FROM   Usuario u
                JOIN   Rol     r ON u.Rol_ID    = r.ID
                LEFT JOIN Ciudad c ON u.Ciudad_ID = c.ID
                LEFT JOIN Pais   p ON c.Pais_ID   = p.ID
                ORDER BY u.ID
                """;

        // Mapea cada fila al DTO administrativo, manejando fecha nula
        return DatabaseManager.executeQuery(sql, rs -> {
            UsuarioAdminDTO dto = new UsuarioAdminDTO();
            dto.setId(rs.getInt("ID"));
            dto.setUsername(rs.getString("Username"));
            dto.setNombre(rs.getString("Nombre"));
            dto.setApellido(rs.getString("Apellido"));
            dto.setCorreo(rs.getString("Correo"));
            dto.setTelefono(rs.getString("Telefono"));

            java.sql.Date fecha = rs.getDate("Fecha_Nacimiento");
            dto.setFechaNacimiento(fecha != null ? fecha.toString() : null);

            dto.setRolId(rs.getInt("Rol_ID"));
            dto.setRolNombre(rs.getString("RolNombre"));
            dto.setCiudad(rs.getString("Ciudad"));
            dto.setPais(rs.getString("Pais"));
            return dto;
        });
    }

    /**
     * Actualiza el rol asignado a un usuario.
     * @param usuarioId  ID del usuario a modificar.
     * @param nuevoRolId ID del nuevo rol a asignar.
     */
    public void actualizarRol(int usuarioId, int nuevoRolId) {
        String sql = "UPDATE Usuario SET Rol_ID = ? WHERE ID = ?";
        DatabaseManager.executeUpdate(sql, nuevoRolId, usuarioId);
    }

    // ── Actualizaciones de perfil editables ────────────────────────────────────

    /** Actualiza nombre, apellido y fecha de nacimiento del usuario. */
    public void actualizarDatosPersonales(int usuarioId, String nombre, String apellido, java.sql.Date fechaNacimiento) {
        String sql = "UPDATE Usuario SET Nombre = ?, Apellido = ?, Fecha_Nacimiento = ? WHERE ID = ?";
        DatabaseManager.executeUpdate(sql, nombre, apellido, fechaNacimiento, usuarioId);
    }

    /** Actualiza el username del usuario. */
    public void actualizarUsername(int usuarioId, String username) {
        String sql = "UPDATE Usuario SET Username = ? WHERE ID = ?";
        DatabaseManager.executeUpdate(sql, username, usuarioId);
    }

    /** Actualiza el correo del usuario. */
    public void actualizarCorreo(int usuarioId, String correo) {
        String sql = "UPDATE Usuario SET Correo = ? WHERE ID = ?";
        DatabaseManager.executeUpdate(sql, correo, usuarioId);
    }

    /** Actualiza el pasaporte del usuario. */
    public void actualizarPasaporte(int usuarioId, String pasaporte) {
        String sql = "UPDATE Usuario SET Pasaporte = ? WHERE ID = ?";
        DatabaseManager.executeUpdate(sql, pasaporte, usuarioId);
    }

    /** Actualiza la ciudad del usuario. */
    public void actualizarCiudad(int usuarioId, int ciudadId) {
        String sql = "UPDATE Usuario SET Ciudad_ID = ? WHERE ID = ?";
        DatabaseManager.executeUpdate(sql, ciudadId, usuarioId);
    }

    /** Actualiza las preferencias de ofertas del usuario (JSON string o null). */
    public void actualizarPreferencias(int usuarioId, String preferenciasOferta) {
        String sql = "UPDATE Usuario SET Preferencias_Oferta = ? WHERE ID = ?";
        DatabaseManager.executeUpdate(sql, preferenciasOferta, usuarioId);
    }
}